enhancements/dns/forward-dns-over-tls.md

```diff
@@ -0,0 +1,266 @@
+---
+title: enable-the-ability-to-configure-dns-over-tls-for-forwarded-requests
+authors:
+  - "@brandisher"
+reviewers:
+  - "@Miciah"
+  - "@rfredette"
+  - "@candita"
+  - "@frobware"
+  - "@knobunc"
+approvers:
+  - "@Miciah"
+  - "@knobunc"
+creation-date: 2021-12-08
+last-updated: 2021-12-08
+tracking-link:
+  - https://issues.redhat.com/browse/NE-750
+see-also:
+  - "/enhancements/dns/plugins.md"
+---
+
+# Enable the ability to configure DNS-over-TLS for forwarded requests
+
+## Summary
+
+This enhancement adds optional configuration to enable DNS-over-TLS when 
+either `spec.servers[].forwardPlugin` and/or `spec.upstreamResolvers` are 
+configured in the cluster DNS spec.
+
+## Motivation
+
+In environments that are highly regulated or geographically dispersed there
+may be a need to forward requests to upstream DNS resolvers. Because DNS is
+unencrypted by default there are concerns in these environments around data
+privacy and integrity. Currently, there is no way to configure DNS-over-TLS for
+upstream resolvers so there is also no workaround to alleviate this security
+concern. The expansion of the `UpstreamResolver` and `ForwardPlugin` APIs  for 
+DNS request forwarding improves the current security posture for DNS on 
+OpenShift.
+
+### Goals
+
+1. A cluster administrator can specify a CA certificate or CA bundle, TLS configuration, and an upstream FQDN for the default zone in the DNS configuration.
+2. A cluster administrator can specify a CA certificate or CA bundle, TLS configuration, and an upstream FQDN for a non-default zone.
+3. A cluster administrator can opt-in to DNS-over-TLS (using the configured CA/bundle) for each listed upstream in the default zone.
+4. A cluster administrator can opt-in to DNS-over-TLS (using the configured CA/bundle) for a set of upstreams on a non-default zone.
+
+### Non-Goals
+
+* Allowing configuration of DNS-over-TLS for use cases other than OpenShift's default DNS provider to upstream resolvers is out of scope.
+
+## Proposal
+
+This proposal covers changes to two APIs. The first is a change to 
+`UpstreamResolvers`.  To enable a cluster administrator to configure 
+DNS-over-TLS, the `UpstreamResolvers` type will get three new optional members:
+
+`TLSSecurityProfile` encapsulates TLS specific configuration. Including this 
+allows customers to set their desired TLS version, ciphers, etc.
+
+`CABundle` references a ConfigMap that contains a single CA certificate 
+or a CA bundle.
+
+`ServerName` contains an FQDN to validate when establishing the connection. 
+This is relevant in particular with cases where a customer has multiple 
+upstreams behind one serving certificate (e.g. `1.1.1.1` and `2.2.2.2` 
+behind `dns.myorg.com`)
+
+Since `UpstreamResolvers` doesn't allow for configuration of a custom zone, 
+it's assumed that the same TLS configuration will apply to _n_ `Upstream` 
+objects. 
```

Comment:
I think we're at an early enough stage that the design could be adjusted pending your approval. I've kept `upstreamResolvers` consistent in heirarchy with `forwardPlugin` in that you can define a cabundle and servername at the top level forwarding/upstream block. However, since Upstream is a little more flexible we could conceivably allow cabundle and servername to live at the `Upstream` level. That would eliminate the need for the `EnableTLS` field and answer some functional questions.

Example:
```
spec:
  upstreamResolvers:
    upstreams:
    - type: SystemResolvConf
      address: 1.1.1.1
      port: 5353
      caBundle: dns-tls-configmap
      servername: dns.bar.com
```
Then the assumption would be that all of the resolvers in `/etc/resolv.conf` would use TLS mostly because there's no way to split those out further.

What do you see as the action items on this one? Should this be a design change or a godoc change?

Reply:
Is it a requirement that it be possible to enable TLS for nameservers that are specified in `/etc/resolv.conf`?  I'm gradually coming around to two realizations:

* For `type: SystemResolvConf`, we really shouldn't deviate from `/etc/resolv.conf` semantics because `/etc/resolv.conf` is coming from the node host's `/etc/resolv.conf`, and in that context, the nameservers are non-TLS, so we should assume that they are non-TLS in general.  
* `Upstream` probably should have been a union type.

Reply:
I've reached out to PM about this one also; I want to get clarity on your question regarding nameservers in `/etc/resolv.conf`.

I could use some direction on where to go with this item next. From what I understand, there's probably a use case for defining `CABundle` and `ServerName` in the `Upstream` instead of in `UpstreamResolvers` since it may not apply to all of the upstreams.

The other action item is addressing the `SystemResolvConf` case. That would have to be handled by a conditional statement in the implementation, correct? I don't see anything in kubebuilder that allows for conditional definition on a field.

Reply:
Correct.  I'm thinking it might have been better to have made the `Upstream` type that we use for `spec.upstreamResolvers.upstreams[*]` a union type, as defined in <https://github.com/kubernetes/enhancements/blob/0e4d5df19d396511fe41ed0860b0ab9b96f46a2d/keps/sig-api-machinery/1027-api-unions/README.md>.  That would have enabled us to specify a group of parameters that could be specified specifically when `spec.upstreamResolvers.upstreams[*].type` were set to "Network".  With feature freeze approaching, though, I don't want to go and try to revise the API now.  So we really just need to leave it up to the operator to ignore fields that are inapplicable for the specified type.  

That aside, in the Corefile, a server block can have a `forward` block with a list of upstreams (which may include any combination of `/etc/resolv.conf` and explicit addresses) andparameters for the TLS cert and server name, and it is better for the API not to allow the user to express something that cannot be expressed in the Corefile, so I think `caBundle` and `servername` belong under `upstreamResolvers`, not under `upstreams`.  We can keep `enableTLS`, which would belong under `upstreams` but would be inapplicable for `SystemResolvConf`.